### PR TITLE
feat: add `influxdb_iox debug build-catalog` command

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/7744

# Rationale
TLDR is I want to reproduce issues using production data locally with all in one mode. 

See https://github.com/influxdata/influxdb_iox/issues/7744 for more details

# Changes

Implement `influxdb_iox debug build-catalog` that takes the output of `influxdb_iox remote get-table` and creates a directory with a catalog and object_store that can be run locally via `influxdb_iox run --data-dir <data_dir>`

# Docs
```
Build a local catalog from the output of `remote get-table`. For example:

 # download contents of table_name into a directory named 'table_name'
 influxdb_iox remote get-table <namespace> <table_name>

 # Create a catalog and object_store in /tmp/data_dir
 influxdb_iox debug build-catalog <table_dir> /tmp/data_dir

 # Start iox using this data directory (you can now query `table_name` locally):
 influxdb_iox --data-dir /tmp/data_dir

Usage: influxdb_iox debug build-catalog [OPTIONS] <INPUT_DIR> <DATA_DIR>

Arguments:
  <INPUT_DIR>
          Directory containing the output of running `influxdb_iox remote store get-table`

  <DATA_DIR>
          Target data directory to create a sqlite catalog and file object_store.

          After a successful rebuild, you can examine the catalog locally using `influxdb_iox --data-dir <dir>`.
```

# Demonstration

## Setup
Running `influxdb_iox` and getting local telegraph data

```shell
(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ influxdb_iox namespace list
[
  {
    "id": "1",
    "name": "26f7e5a4b7be365b_917b97a92e883afc",
    "maxTables": 500,
    "maxColumnsPerTable": 200
  }
]
```

## Export `cpu` table:
```shell
(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ influxdb_iox remote store get-table 26f7e5a4b7be365b_917b97a92e883afc cpu
found 11 Parquet files, exporting...
downloading file 1 of 11 (1b8eb36a-7a34-4635-9156-251efcb1c024.4.parquet)...
downloading file 2 of 11 (1819137f-7cb5-4dc8-8051-6fa0b42990cb.4.parquet)...
downloading file 3 of 11 (4931cad7-7aaf-4b41-8f46-2d3be85c492b.4.parquet)...
downloading file 4 of 11 (be75f5fb-a8bc-4646-893a-70d496b13f3d.4.parquet)...
downloading file 5 of 11 (5235b87d-19ee-48ae-830f-b19d81bfe915.4.parquet)...
downloading file 6 of 11 (a8f7be33-42b6-4353-8735-51b245196d39.4.parquet)...
downloading file 7 of 11 (3b43c4ee-7500-47f9-9c0f-76d4f80b480e.4.parquet)...
downloading file 8 of 11 (081da5be-e0f9-4b42-8cd2-45bfebbd934c.4.parquet)...
downloading file 9 of 11 (f29ba3b4-53b1-4c68-9287-4bcea7c4e86b.4.parquet)...
downloading file 10 of 11 (1ce94ce2-1200-4516-950a-64828a7cebba.4.parquet)...
downloading file 11 of 11 (3a2b5525-3be5-41ef-b082-b279edc32acb.4.parquet)...
Done.
(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ ls cpu/
081da5be-e0f9-4b42-8cd2-45bfebbd934c.4.parquet       1ce94ce2-1200-4516-950a-64828a7cebba.4.parquet       4931cad7-7aaf-4b41-8f46-2d3be85c492b.4.parquet       be75f5fb-a8bc-4646-893a-70d496b13f3d.4.parquet
081da5be-e0f9-4b42-8cd2-45bfebbd934c.4.parquet.json  1ce94ce2-1200-4516-950a-64828a7cebba.4.parquet.json  4931cad7-7aaf-4b41-8f46-2d3be85c492b.4.parquet.json  be75f5fb-a8bc-4646-893a-70d496b13f3d.4.parquet.json
1819137f-7cb5-4dc8-8051-6fa0b42990cb.4.parquet       3a2b5525-3be5-41ef-b082-b279edc32acb.4.parquet       5235b87d-19ee-48ae-830f-b19d81bfe915.4.parquet       f29ba3b4-53b1-4c68-9287-4bcea7c4e86b.4.parquet
1819137f-7cb5-4dc8-8051-6fa0b42990cb.4.parquet.json  3a2b5525-3be5-41ef-b082-b279edc32acb.4.parquet.json  5235b87d-19ee-48ae-830f-b19d81bfe915.4.parquet.json  f29ba3b4-53b1-4c68-9287-4bcea7c4e86b.4.parquet.json
1b8eb36a-7a34-4635-9156-251efcb1c024.4.parquet       3b43c4ee-7500-47f9-9c0f-76d4f80b480e.4.parquet       a8f7be33-42b6-4353-8735-51b245196d39.4.parquet       partition.4.json
1b8eb36a-7a34-4635-9156-251efcb1c024.4.parquet.json  3b43c4ee-7500-47f9-9c0f-76d4f80b480e.4.parquet.json  a8f7be33-42b6-4353-8735-51b245196d39.4.parquet.json  table.1.json
```

## Build a new `new_data_dir` from export:
```shell
(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ influxdb_iox debug build-catalog cpu new_data_dir
Beginning catalog / object_store build from "cpu" in "new_data_dir"....
Done

(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ ls new_data_dir/
catalog.sqlite  object_store/
```

## Run `influxdb_iox` with `new_data_dir`:

```shell
(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ influxdb_iox --data-dir new_data_dir/
```

And in a separate shell, you can query the data and see it is present:

```shell
(arrow_dev) alamb@MacBook-Pro-8:~/Software/influxdb_iox2$ influxdb_iox query 26f7e5a4b7be365b_917b97a92e883afc 'select * from cpu limit 10';
+-----------+---------------------+----------------------+-------------+------------------+-------------------+--------------+-----------+------------+---------------+-------------+--------------------+--------------------+
| cpu       | host                | time                 | usage_guest | usage_guest_nice | usage_idle        | usage_iowait | usage_irq | usage_nice | usage_softirq | usage_steal | usage_system       | usage_user         |
+-----------+---------------------+----------------------+-------------+------------------+-------------------+--------------+-----------+------------+---------------+-------------+--------------------+--------------------+
| cpu-total | MacBook-Pro-8.local | 2023-07-06T17:13:40Z | 0.0         | 0.0              | 95.6668753914105  | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 1.4902943018170824 | 2.8428303068453085 |
| cpu-total | MacBook-Pro-8.local | 2023-07-06T17:13:50Z | 0.0         | 0.0              | 95.9551687433697  | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 1.4213261536472683 | 2.6235051029648098 |
| cpu-total | MacBook-Pro-8.local | 2023-07-06T17:14:00Z | 0.0         | 0.0              | 96.52108622167991 | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 1.37029157802418   | 2.108622199968126  |
| cpu-total | MacBook-Pro-8.local | 2023-07-06T17:14:10Z | 0.0         | 0.0              | 95.26819803491809 | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 1.752519246414341  | 2.979282718922596  |
| cpu-total | MacBook-Pro-8.local | 2023-07-06T17:14:20Z | 0.0         | 0.0              | 95.28402329791422 | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 1.6408843239063593 | 3.0750923780335997 |
| cpu-total | MacBook-Pro-8.local | 2023-07-06T17:14:30Z | 0.0         | 0.0              | 93.97484827633119 | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 2.0271538509716924 | 3.9979978727699588 |
| cpu-total | MacBook-Pro-8.local | 2023-07-06T17:14:40Z | 0.0         | 0.0              | 95.69219209824692 | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 1.458894245831095  | 2.848913656031324  |
| cpu-total | MacBook-Pro-8.local | 2023-07-06T17:14:50Z | 0.0         | 0.0              | 94.78402607970591 | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 1.9685286188771443 | 3.2474453011797517 |
| cpu-total | MacBook-Pro-8.local | 2023-07-06T17:15:00Z | 0.0         | 0.0              | 95.85132344665212 | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 1.5706151054475623 | 2.5780614479731607 |
| cpu0      | MacBook-Pro-8.local | 2023-07-06T17:13:40Z | 0.0         | 0.0              | 78.65055387717186 | 0.0          | 0.0       | 0.0        | 0.0           | 0.0         | 7.452165156077374  | 13.897280966824042 |
+-----------+---------------------+----------------------+-------------+------------------+-------------------+--------------+-----------+------------+---------------+-------------+--------------------+--------------------+
```

### CODE REVIEW

File: influxdb_iox/src/commands/debug/build_catalog.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+use object_store::ObjectStore;
+use observability_deps::tracing::info;
+use snafu::{ResultExt, Snafu};
+use std::{
+    io::Write,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+#[derive(Debug, Snafu)]
```
Comment (alamb) [inline influxdb_iox/src/commands/debug/build_catalog.rs:14]: This is the wiring into the command line

File: influxdb_iox/tests/end_to_end_cases/debug.rs [hunk 2]
```diff
@@ -27,3 +38,175 @@ async fn test_print_cpu() {
             "rustc is using the following target options",
         ));
 }
+
+/// Tests that we can
```
Comment (alamb) [inline influxdb_iox/tests/end_to_end_cases/debug.rs:42]: This is an end to end test (rather than a unit test) that ensures the end to end flow works, as discussed with @crepererum 

It isn't perfect coverage, but I think it is acceptable for a debug tool

File: import_export/src/file/import.rs [hunk 3]
```diff
@@ -92,9 +164,51 @@ impl ExportedContents {
...
+                e,
+            })?;
+
+            let parquet_file: proto::ParquetFile =
+                serde_json::from_str(&json).map_err(|e| Error::Json {
+                    path: path.clone(),
+                    e,
+                })?;
+
+            //info!(?path, object_store_id=%parquet_file.object_store_id, "Read parquet metadata");
```
Comment (crepererum) [inline import_export/src/file/import.rs:205]: leftover code?

File: import_export/src/file/import.rs [hunk 4]
```diff
@@ -124,3 +259,440 @@ fn file_name(p: &Path) -> Cow<'_, str> {
...
+        iox_metadata: &IoxMetadata,
+        decoded_iox_parquet_metadata: &DecodedIoxParquetMetaData,
+        file_size_bytes: usize,
+    ) -> Result<ParquetFileParams> {
+        let object_store_id = iox_metadata.object_store_id;
+
+        // need to make columns in the target catalog
+        let column_set = insert_columns(table.id, decoded_iox_parquet_metadata, repos).await?;
+
+        // TODO: do we need to update the partition_hash_id?
```
Comment (crepererum) [inline import_export/src/file/import.rs:571]: Yes, might be generated from `partition_key` if you have that at hand.

**Review by crepererum — APPROVED** [review]
Some comments, overall a very nice tool to have :+1:

File: influxdb_iox/src/commands/debug/build_catalog.rs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        path: &os_dir,
+    })?;
+
+    info!(?os_dir, "Using local object store");
+    let object_store = object_store::local::LocalFileSystem::new_with_prefix(os_dir)?;
+
+    Ok(Arc::new(object_store))
+}
+
+struct StdOutObserver {}
```
[the remark this answers is not in the document]
Comment (alamb) [reply influxdb_iox/src/commands/debug/build_catalog.rs]: Done in 3ffb3427e

File: influxdb_iox/src/commands/debug/build_catalog.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    let catalog = get_catalog(&data_dir).await?;
+    catalog.setup().await?;
+
+    let object_store = get_object_store(&data_dir)?;
+
+    info!("Initialized catalog, object store, and input path ...");
+
+    let importer = RemoteImporter::new(exported_contents, catalog, object_store);
+
+    println!("Beginning catalog / object_store build from {input_dir:?} in {data_dir:?}....");
```
[the remark this answers is not in the document]
Comment (alamb) [reply influxdb_iox/src/commands/debug/build_catalog.rs]: In 3ffb3427e

File: import_export/src/file/import.rs (hunk 4 above)
Comment (alamb) [reply import_export/src/file/import.rs]: in d36b583cfeb3ff8f8433397870aa7d74196510bb
